File: kernel/db/btree.c
#include "btree.h"
#include "../mm/heap.h"
#include "../lib/string.h"
#include "../lib/assert.h"

static btree_node_t *btree_create_node(bool is_leaf) {
    btree_node_t *node = kzalloc(sizeof(btree_node_t));
    ASSERT(node != NULL);
    node->is_leaf = is_leaf;
    node->num_keys = 0;
    return node;
}

void btree_init(btree_t *tree, uint32_t table_id) {
    tree->root = btree_create_node(true);
    tree->count = 0;
    tree->table_id = table_id;
}

static void *btree_search_node(btree_node_t *node, uint64_t key) {
    if (!node) return NULL;

    uint32_t i = 0;
    while (i < node->num_keys && key > node->keys[i]) i++;

    if (i < node->num_keys && key == node->keys[i]) {
        return node->values[i];
    }

    if (node->is_leaf) return NULL;
    return btree_search_node(node->children[i], key);
}

void *btree_search(btree_t *tree, uint64_t key) {
    return btree_search_node(tree->root, key);
}

static void btree_split_child(btree_node_t *parent, uint32_t index) {
    btree_node_t *child = parent->children[index];
    uint32_t mid = BTREE_MAX_KEYS / 2;

    btree_node_t *new_node = btree_create_node(child->is_leaf);
    new_node->num_keys = child->num_keys - mid - 1;

    /* Copy upper half of keys/values to new node */
    for (uint32_t j = 0; j < new_node->num_keys; j++) {
        new_node->keys[j] = child->keys[mid + 1 + j];
        new_node->values[j] = child->values[mid + 1 + j];
    }

    if (!child->is_leaf) {
        for (uint32_t j = 0; j <= new_node->num_keys; j++) {
            new_node->children[j] = child->children[mid + 1 + j];
        }
    }

    /* Shift parent's keys/children right */
    for (int j = (int)parent->num_keys; j > (int)index; j--) {
        parent->keys[j] = parent->keys[j - 1];
        parent->values[j] = parent->values[j - 1];
        parent->children[j + 1] = parent->children[j];
    }

    /* Insert median key into parent */
    parent->keys[index] = child->keys[mid];
    parent->values[index] = child->values[mid];
    parent->children[index + 1] = new_node;
    parent->num_keys++;

    child->num_keys = mid;
}

static void btree_insert_nonfull(btree_node_t *node, uint64_t key, void *value) {
    int i = (int)node->num_keys - 1;

    if (node->is_leaf) {
        /* Shift keys right to make room */
        while (i >= 0 && key < node->keys[i]) {
            node->keys[i + 1] = node->keys[i];
            node->values[i + 1] = node->values[i];
            i--;
        }
        /* Check for duplicate key (update in place) */
        if (i >= 0 && node->keys[i] == key) {
            node->values[i] = value;
            return;
        }
        node->keys[i + 1] = key;
        node->values[i + 1] = value;
        node->num_keys++;
    } else {
        while (i >= 0 && key < node->keys[i]) i--;
        if (i >= 0 && node->keys[i] == key) {
            node->values[i] = value;
            return;
        }
        i++;
        if (node->children[i]->num_keys == BTREE_MAX_KEYS) {
            btree_split_child(node, (uint32_t)i);
            if (key > node->keys[i]) i++;
            if (key == node->keys[i]) {
                node->values[i] = value;
                return;
            }
        }
        btree_insert_nonfull(node->children[i], key, value);
    }
}

int btree_insert(btree_t *tree, uint64_t key, void *value) {
    btree_node_t *root = tree->root;

    if (root->num_keys == BTREE_MAX_KEYS) {
        btree_node_t *new_root = btree_create_node(false);
        new_root->children[0] = root;
        btree_split_child(new_root, 0);
        tree->root = new_root;

        int i = (key > new_root->keys[0]) ? 1 : 0;
        if (key == new_root->keys[0]) {
            new_root->values[0] = value;
        } else {
            btree_insert_nonfull(new_root->children[i], key, value);
        }
    } else {
        btree_insert_nonfull(root, key, value);
    }

    tree->count++;
    return 0;
}

/* Simple delete: mark as NULL (lazy deletion for MVP) */
int btree_delete(btree_t *tree, uint64_t key) {
    /* Find and remove from leaf (simplified) */
    btree_node_t *node = tree->root;
    while (node) {
        uint32_t i = 0;
        while (i < node->num_keys && key > node->keys[i]) i++;

        if (i < node->num_keys && key == node->keys[i]) {
            if (node->is_leaf) {
                /* Shift keys left */
                for (uint32_t j = i; j < node->num_keys - 1; j++) {
                    node->keys[j] = node->keys[j + 1];
                    node->values[j] = node->values[j + 1];
                }
                node->num_keys--;
                tree->count--;
                return 0;
            }
            /* For internal nodes in MVP: mark value as NULL */
            node->values[i] = NULL;
            tree->count--;
            return 0;
        }

        if (node->is_leaf) return -1; /* Not found */
        node = node->children[i];
    }
    return -1;
}

static void btree_scan_node(btree_node_t *node, btree_iter_fn callback, void *ctx) {
    if (!node) return;

    for (uint32_t i = 0; i < node->num_keys; i++) {
        if (!node->is_leaf) {
            btree_scan_node(node->children[i], callback, ctx);
        }
        if (node->values[i]) {
            callback(node->keys[i], node->values[i], ctx);
        }
    }
    if (!node->is_leaf) {
        btree_scan_node(node->children[node->num_keys], callback, ctx);
    }
}

void btree_scan(btree_t *tree, btree_iter_fn callback, void *ctx) {
    btree_scan_node(tree->root, callback, ctx);
}

static void btree_destroy_node(btree_node_t *node) {
    if (!node) return;
    if (!node->is_leaf) {
        for (uint32_t i = 0; i <= node->num_keys; i++) {
            btree_destroy_node(node->children[i]);
        }
    }
    kfree(node);
}

void btree_destroy(btree_t *tree) {
    btree_destroy_node(tree->root);
    tree->root = NULL;
    tree->count = 0;
}

File: vaultos-rs/kernel/src/db/btree.rs
// Order-64 B-tree for VaultOS-RS
// Direct port of kernel/db/btree.c

use alloc::boxed::Box;

pub const BTREE_ORDER: usize = 64;
pub const BTREE_MAX_KEYS: usize = BTREE_ORDER - 1;
pub const BTREE_MIN_KEYS: usize = BTREE_ORDER / 2 - 1;

pub struct BtreeNode {
    pub keys: [u64; BTREE_MAX_KEYS],
    pub values: [*mut u8; BTREE_MAX_KEYS],   // void* equivalent
    pub children: [*mut BtreeNode; BTREE_ORDER],
    pub num_keys: u32,
    pub is_leaf: bool,
    // Persistence
    pub disk_lba: u64,
    pub value_lbas: [u64; BTREE_MAX_KEYS],
    pub child_lbas: [u64; BTREE_ORDER],
    pub dirty: bool,
}

unsafe impl Send for BtreeNode {}
unsafe impl Sync for BtreeNode {}

impl BtreeNode {
    pub fn new(is_leaf: bool) -> *mut BtreeNode {
        let node = Box::new(BtreeNode {
            keys: [0u64; BTREE_MAX_KEYS],
            values: [core::ptr::null_mut(); BTREE_MAX_KEYS],
            children: [core::ptr::null_mut(); BTREE_ORDER],
            num_keys: 0,
            is_leaf,
            disk_lba: 0,
            value_lbas: [0u64; BTREE_MAX_KEYS],
            child_lbas: [0u64; BTREE_ORDER],
            dirty: true,
        });
        Box::into_raw(node)
    }
}

pub struct Btree {
    pub root: *mut BtreeNode,
    pub count: u64,
    pub table_id: u32,
}

unsafe impl Send for Btree {}
unsafe impl Sync for Btree {}

pub type BtreeIterFn = fn(u64, *mut u8, *mut u8);

pub fn btree_init(tree: &mut Btree, table_id: u32) {
    tree.root = BtreeNode::new(true);
    tree.count = 0;
    tree.table_id = table_id;
}

pub fn btree_search(tree: &Btree, key: u64) -> *mut u8 {
    unsafe { btree_search_node(tree.root, key) }
}

unsafe fn btree_search_node(node: *mut BtreeNode, key: u64) -> *mut u8 {
    if node.is_null() { return core::ptr::null_mut(); }

    let n = &*node;
    let mut i = 0u32;
    while i < n.num_keys && key > n.keys[i as usize] { i += 1; }

    if i < n.num_keys && key == n.keys[i as usize] {
        return n.values[i as usize];
    }

    if n.is_leaf { return core::ptr::null_mut(); }
    btree_search_node(n.children[i as usize], key)
}

unsafe fn btree_split_child(parent: *mut BtreeNode, index: u32) {
    let p = &mut *parent;
    // Guard: parent must not be full — caller invariant.  Prevents children[64] OOB.
    if p.num_keys >= BTREE_MAX_KEYS as u32 {
        return;
    }
    let child = &mut *p.children[index as usize];
    let mid = (BTREE_MAX_KEYS / 2) as u32;

    let new_node = BtreeNode::new(child.is_leaf);
    let nn = &mut *new_node;
    nn.num_keys = child.num_keys - mid - 1;

    // Copy upper half to new node
    for j in 0..nn.num_keys {
        nn.keys[j as usize] = child.keys[(mid + 1 + j) as usize];
        nn.values[j as usize] = child.values[(mid + 1 + j) as usize];
        nn.value_lbas[j as usize] = child.value_lbas[(mid + 1 + j) as usize];
    }

    if !child.is_leaf {
        for j in 0..=nn.num_keys {
            nn.children[j as usize] = child.children[(mid + 1 + j) as usize];
            nn.child_lbas[j as usize] = child.child_lbas[(mid + 1 + j) as usize];
        }
    }

    // Shift parent's keys/children right
    let mut j = p.num_keys as i32;
    while j > index as i32 {
        p.keys[j as usize] = p.keys[(j - 1) as usize];
        p.values[j as usize] = p.values[(j - 1) as usize];
        p.value_lbas[j as usize] = p.value_lbas[(j - 1) as usize];
        p.children[(j + 1) as usize] = p.children[j as usize];
        p.child_lbas[(j + 1) as usize] = p.child_lbas[j as usize];
        j -= 1;
    }

    // Insert median into parent
    p.keys[index as usize] = child.keys[mid as usize];
    p.values[index as usize] = child.values[mid as usize];
    p.value_lbas[index as usize] = child.value_lbas[mid as usize];
    p.children[(index + 1) as usize] = new_node;
    p.child_lbas[(index + 1) as usize] = 0;
    p.num_keys += 1;

    child.num_keys = mid;

    p.dirty = true;
    child.dirty = true;
    nn.dirty = true;
}

unsafe fn btree_insert_nonfull(node: *mut BtreeNode, key: u64, value: *mut u8) {
    let n = &mut *node;
    let mut i = n.num_keys as i32 - 1;

    if n.is_leaf {
        while i >= 0 && key < n.keys[i as usize] {
            n.keys[(i + 1) as usize] = n.keys[i as usize];
            n.values[(i + 1) as usize] = n.values[i as usize];
            n.value_lbas[(i + 1) as usize] = n.value_lbas[i as usize];
            i -= 1;
        }
        // Check for duplicate key (update in place)
        if i >= 0 && n.keys[i as usize] == key {
            n.values[i as usize] = value;
            n.value_lbas[i as usize] = 0;
            n.dirty = true;
            return;
        }
        n.keys[(i + 1) as usize] = key;
        n.values[(i + 1) as usize] = value;
        n.value_lbas[(i + 1) as usize] = 0;
        n.num_keys += 1;
        n.dirty = true;
    } else {
        while i >= 0 && key < n.keys[i as usize] { i -= 1; }
        if i >= 0 && n.keys[i as usize] == key {
            n.values[i as usize] = value;
            n.value_lbas[i as usize] = 0;
            n.dirty = true;
            return;
        }
        i += 1;
        if (*n.children[i as usize]).num_keys == BTREE_MAX_KEYS as u32 {
            btree_split_child(node, i as u32);
            if key > n.keys[i as usize] { i += 1; }
            if key == n.keys[i as usize] {
                n.values[i as usize] = value;
                n.value_lbas[i as usize] = 0;
                n.dirty = true;
                return;
            }
        }
        btree_insert_nonfull(n.children[i as usize], key, value);
    }
}

pub fn btree_insert(tree: &mut Btree, key: u64, value: *mut u8) -> i32 {
    unsafe {
        let root = tree.root;
        if (*root).num_keys == BTREE_MAX_KEYS as u32 {
            let new_root = BtreeNode::new(false);
            (*new_root).children[0] = root;
            (*new_root).child_lbas[0] = (*root).disk_lba;
            btree_split_child(new_root, 0);
            tree.root = new_root;

            let i = if key > (*new_root).keys[0] { 1 } else { 0 };
            if key == (*new_root).keys[0] {
                (*new_root).values[0] = value;
            } else {
                btree_insert_nonfull((*new_root).children[i], key, value);
            }
        } else {
            btree_insert_nonfull(root, key, value);
        }
    }
    tree.count += 1;
    0
}

pub fn btree_delete(tree: &mut Btree, key: u64) -> i32 {
    unsafe {
        let mut node = tree.root;
        while !node.is_null() {
            let n = &mut *node;
            let mut i = 0u32;
            while i < n.num_keys && key > n.keys[i as usize] { i += 1; }

            if i < n.num_keys && key == n.keys[i as usize] {
                if n.is_leaf {
                    for j in i..(n.num_keys - 1) {
                        n.keys[j as usize] = n.keys[(j + 1) as usize];
                        n.values[j as usize] = n.values[(j + 1) as usize];
                        n.value_lbas[j as usize] = n.value_lbas[(j + 1) as usize];
                    }
                    n.num_keys -= 1;
                    n.dirty = true;
                    tree.count -= 1;
                    return 0;
                }
                // Internal node: mark value NULL (lazy)
                n.values[i as usize] = core::ptr::null_mut();
                n.value_lbas[i as usize] = 0;
                n.dirty = true;
                tree.count -= 1;
                return 0;
            }

            if n.is_leaf { return -1; }
            node = n.children[i as usize];
        }
    }
    -1
}

unsafe fn btree_scan_node(node: *mut BtreeNode, callback: BtreeIterFn, ctx: *mut u8) {
    if node.is_null() { return; }
    let n = &*node;

    for i in 0..n.num_keys as usize {
        if !n.is_leaf {
            btree_scan_node(n.children[i], callback, ctx);
        }
        if !n.values[i].is_null() {
            callback(n.keys[i], n.values[i], ctx);
        }
    }
    if !n.is_leaf {
        btree_scan_node(n.children[n.num_keys as usize], callback, ctx);
    }
}

pub fn btree_scan(tree: &Btree, callback: BtreeIterFn, ctx: *mut u8) {
    unsafe { btree_scan_node(tree.root, callback, ctx); }
}

unsafe fn btree_destroy_node(node: *mut BtreeNode) {
    if node.is_null() { return; }
    let n = &*node;
    if !n.is_leaf {
        for i in 0..=n.num_keys as usize {
            btree_destroy_node(n.children[i]);
        }
    }
    let _ = Box::from_raw(node);
}

pub fn btree_destroy(tree: &mut Btree) {
    unsafe { btree_destroy_node(tree.root); }
    tree.root = core::ptr::null_mut();
    tree.count = 0;
}
